### ISSUE TITLE
Infinate querying when query fails and using suspense + enabled flag

### ISSUE THREAD

**Issue by vimcaw** [tannerlinsley/react-query#2434]
**Describe the bug**
Using suspense + enabled flag to delay querying, when a query fails will cause infinite querying.

**To Reproduce**

Core code:
```ts
function query() {
  return new Promise((resolve, reject) => {
    console.count("query");
    setTimeout(() => reject("Query Error"), 1000);
  });
}

function DataFetch() {
  const [enabled, setEnabledStatus] = useState(false);
  useQuery("mock", query, { suspense: true, enabled, retry: 0 });

  return <button onClick={() => setEnabledStatus(true)}>Start</button>;
}
```

Complete code: https://codesandbox.io/s/react-query-bug-suspense-enabled-flag-2w762?file=/src/App.tsx

1. Click the `Start` Button
2. Look at `Console`, you will see many queries.

**Expected behavior**
Throw Error and only query once.

**Screenshots**
![image](https://user-images.githubusercontent.com/21964227/124413644-a4c8a800-dd83-11eb-8159-6d27b8a2a7dd.png)

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser: chrome
 - Version: 91.0.4472.114

**Additional context**
After removing the `enabled` or `suspense` flag, anything works well.

**Comment by flybayer** [comment]
We've also ran into this [with Blitz](https://github.com/blitz-js/blitz/issues/2538).

Here's a failing test to add to [suspense.test.tsx](https://github.com/tannerlinsley/react-query/blob/master/src/react/tests/suspense.test.tsx): 
```ts
  it('should not have infinite loop with error query when enabled goes from false to true', async () => {
    const key = queryKey()

    const consoleMock = mockConsoleError()

    let queryExecutions = 0

    function Page() {
      const [count, setCount] = React.useState(0)
      useQuery(
        key,
        async () => {
          console.log('Running query...')
          queryExecutions++
          await sleep(10)
          throw new Error('Suspense Error')
        },
        {
          enabled: count > 0,
          retry: false,
          suspense: true,
        }
      )
      return (
        <div>
          rendered
          <button onClick={() => setCount(c => c + 1)}>increment</button>
        </div>
      )
    }

    function App() {
      const { reset } = useQueryErrorResetBoundary()
      return (
        <ErrorBoundary
          onReset={reset}
          fallbackRender={() => <div>error boundary</div>}
        >
          <React.Suspense fallback="Loading...">
            <Page />
          </React.Suspense>
        </ErrorBoundary>
      )
    }

    const rendered = renderWithClient(queryClient, <App />)

    await waitFor(() => rendered.getByText('rendered'))
    fireEvent.click(rendered.getByText('increment'))
    await waitFor(() => rendered.getByText('Loading...'))
    await waitFor(() => rendered.getByText('rendered'))
    await sleep(100)
    expect(queryExecutions).toBe(1)

    consoleMock.mockRestore()
  })
```

**Comment by cevr** [comment]
We ran into this as well when we removed a query from the cache inside the devtools. The query is permanently stuck in an infinite loop unless we remove the suspense flag, even if we restart the dev server or clear browser cache

**Comment by monstasat** [comment]
We're also running into the same problem even without `enabled` flag set to `false` initially. Setting `suspense: true` for the `useQuery` hook causes the infinite retry loop.

Noticed this after upgrading from `3.5.12` to `3.18.1`

**Comment by juwonjung-hdj** [comment]
This error happens since version `3.9.0`

**Comment by TkDodo** [comment]
3.9.0 had a rather big re-write to keep side-effects out of the render function: https://github.com/tannerlinsley/react-query/pull/1775

**Comment by TkDodo** [comment]
@flybayer yep, the test is failing. I just wanted to ask if the assertions are really expected:

```
    await waitFor(() => rendered.getByText('rendered'))
    fireEvent.click(rendered.getByText('increment'))
    await waitFor(() => rendered.getByText('Loading...'))
    await waitFor(() => rendered.getByText('rendered'))
```

after clicking the `increment` button, going to `loading` seems correct, but shouldn't we then see the fallback from the `errorBoundary` because the query fails?

changing the last line to:

```diff
-     await waitFor(() => rendered.getByText('rendered'))
+    await waitFor(() => rendered.getByText('error boundary'))
```

results in the error that this cannot be found at all, so it seems to be a problem that the error boundary is not shown at all 🤔 ?

**Comment by flybayer** [comment]
@TkDodo ah yes, it should render the error boundary

**Comment by TkDodo** [comment]
> @TkDodo ah yes, it should render the error boundary

okay, it doesn't do that either :/ I tried to debug this a bit but to no success. Works fine without the `enabled` transition, really a bit strange. I couldn't pinpoint where the infinite re-renders were coming from. Maybe @boschni could help us here 🥺

**Comment by mh-alahdadian** [comment]
Any update on this ? I would be happy to work on it

**Comment by TkDodo** [comment]
@mh-alahdadian yes please go ahead if you want :)

**Comment by mh-alahdadian** [comment]
I have found that logic of `shouldFetchOptionally` is wrong 
```ts
(!options.suspense ||
      query.state.status !== 'error' ||
      prevOptions.enabled === false)
```
As you can see this condition should be false when suspense is true but we let it to fetch query again even though we know that it has error

@TkDodo

**Comment by TkDodo** [comment]
we already have extended the condition quite some times to appease the suspense 😅 :

https://github.com/tannerlinsley/react-query/blob/b84681fdc7d37645cd149f40e98cb18fd8bc393c/src/core/queryObserver.ts#L767-L781

maybe you can show how you think the condition should look like in your opinion?

FYI, we added the check in this PR:
- https://github.com/tannerlinsley/react-query/pull/2556

to fix this issue with suspense:
- https://github.com/tannerlinsley/react-query/issues/2187

this also caused a regression for non-suspense usage:
- https://github.com/tannerlinsley/react-query/issues/2682

which we then fixed by making the condition solely for suspense:
- https://github.com/tannerlinsley/react-query/pull/2686

just want to point out that if we touch that again, we have to make sure to _only_ influence suspense usage and _not_ everything else please, because regressions for non-suspense users are not worth it (suspense is still experimental)

**Comment by mh-alahdadian** [comment]
What if we remove `prevOptions.enabled === false` from the condition? I think that is not needed because we know that state is 'error' .
I removed that one and it fixes this bug and all tests are passed as well. is it enough ?

**Comment by TkDodo** [comment]
sure, I think we added a test case for the other situation as well so that we don't break it again. please add the test from this issue as well so that we can see that it passes.

also tagging @JaeYeopHan - maybe he remembers why this condition was added?

**Comment by JaeYeopHan** [comment]
maybe... this point: https://github.com/tannerlinsley/react-query/pull/2556

but... at that time, condition for preventing infinite query was `prevQuery.state.status !== 'error'`.
and now it is not(`options.suspense`), so I think it would be ok to remove the `enabled === false` option! :)

**Comment by TkDodo** [comment]
> but... at that time, condition for preventing infinite query was prevQuery.state.status !== 'error'.
and now it is not(options.suspense), so I think it would be ok to remove the enabled === false option! :)

yeah, could be. I would trust the tests here. If the new test works and the old test still works, too, I'm happy to merge this :)

**Comment by JaeYeopHan** [comment]
@mh-alahdadian Are you planning to contribute?

**Comment by steevsachs** [comment]
Popping in to say thank you guys so much for resolving this! The fix came just as we had identified this as a prod issue and really saved me a ton of grief : )

**Comment by winner106** [comment]
> 🎉 This issue has been resolved in version 3.34.3 🎉
> 
> The release is available on:
> 
> * [npm package (@latest dist-tag)](https://www.npmjs.com/package/react-query/v/3.34.3)
> * [GitHub release](https://github.com/tannerlinsley/react-query/releases/tag/v3.34.3)
> 
> Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot 📦🚀

@flybayer could you please fix the problem in blitz ? thank you very much.
